src/fileformats/ocd_types.cpp

```diff
@@ -100,6 +108,286 @@ namespace Ocd
+	// - the complete output is in 256 bytes of a local array `icon_v8.bits`,
+	//   limiting the maximum number of dictionary entries.
+	// The compressed data may exceed 256 bytes.
+	IconV8 IconV9::compress() const
+	{
+		IconV8 icon_v8 = {};
+		
+		std::vector<ByteSequence> dict;
+		dict.reserve((icon_v8.length() - 16) * 8 / 9);
+		
+		using CodeStream = CodeStreamImplementation<quint8>;
+		auto stream = CodeStream(icon_v8.bits);
+		stream.seek(16);
+		
+		// Searches for the sequence in the dictionary, and returns the matching
+		// index, or -1 if not found.
+		auto find_code = [&dict](const ByteSequence& sequence)->int {
+			if (sequence.count == 1)
+			{
+				return *sequence.start;
+			}
+			for (quint8 i = 0; i < dict.size(); ++i)
+			{
+				const auto& entry = dict[i];
+				if (entry.count == sequence.count
+				    && 0 == memcmp(entry.start, sequence.start, sequence.count))
+					return CodeStream::FirstDynamicCode + i;
+			}
+			return -1;
+		};
+		
+		// Put the first input byte into the buffer, and loop over the rest.
+		auto buffer = ByteSequence{ bits, 1 };
+		for (auto i = 1u; i < length(); ++i)
+		{
+			// Determine code for current buffer + input
+			auto new_buffer = ByteSequence{ bits + i - buffer.count, buffer.count + 1 };
+			auto code = find_code(new_buffer);
+			if (Q_UNLIKELY(code >= 0x200))
+			{
+				// Code word length exeeds 9 bits.
+				throw std::length_error("Excess data during icon compression");
+			}
+			else if (code >= 0)
+			{
+				// Buffer + input was found in the code word table.
+				buffer = new_buffer;
+			}
+			else
+			{
+				// Buffer + input was not found in code word table.
+				dict.push_back(new_buffer);
+				stream.put(quint16(find_code(buffer)));
+				buffer = { bits + i, 1 };
+			}
+		}
+		
+		stream.put(quint16(find_code(buffer)));
+		
+		if (!stream.atEnd())
+			stream.put(CodeStream::EndOfData);
+		
+		return icon_v8;
+	}
+	
+	
+	// Cf. http://giflib.sourceforge.net/whatsinagif/lzw_image_data.html
+	// In IconV8::bits, compressed data starts at offset 16.
+	// Our implementation leverages the facts that:
+	// - the complete input is in 256 bytes of a local array `bits`,
+	//   limiting the maximum number of dictionary entries.
+	// - the complete output is in local array `icon_v9.bits`, so
+	//   the code word dictionary entries can refer to this data.
+	IconV9 IconV8::uncompress() const
+	{
+		IconV9 icon_v9 = {};
+		
+		std::vector<ByteSequence> dict;
+		dict.reserve((length() - 16) * 8 / 9);
+		
+		using CodeStream = CodeStreamImplementation<const quint8>;
+		auto stream = CodeStream(bits);
+		stream.seek(16);
+		
+		// Initialization: handle first code word.
+		auto pos = 0u;  // index in icon_v9.bits
+		icon_v9.bits[pos++] = quint8(stream.get());
+		auto buffer = ByteSequence{ icon_v9.bits, 1 };  // expansion of previous code word
+		
+		while (!stream.atEnd())
+		{
+			auto code = stream.get();
+			
+			// Handle code word
+			if (code <= CodeStream::LastStaticCode)
```

Comment:
LastStaticCode is 0xff. However, for the icons, we could go down to 124, the maximum valid color index.